netbox/netbox_inventory.py

```diff
@@ -132,53 +182,84 @@ def fetch_inventory(cli_opts):
-    # User Filters
-    # -------------------------------------------------------------------------
+    return apply_filters(device_list, cli_opts)
+
+
+def apply_filters(device_list: list[dict], cli_opts: argparse.Namespace) -> Iterator[dict]:
+    """Apply user-provided filters to the device list.
 
-    # If Caller provided an explicit list of device-roles, then filter the
-    # device list based on those roles before creating the inventory
+    Args:
+        device_list: The list of devices to filter.
+        cli_opts: Parsed command-line options.
 
+    Returns:
+        Iterator[dict]: The filtered list of devices.
+    """
     filter_functions = []
 
     if cli_opts.role:
 
-        def filter_role(dev_dict):
+        def filter_role(dev_dict: dict) -> bool:
+            """Filter to include devices that match the specified roles.
+
+            Args:
+                dev_dict (dict): The device record to filter.
+
+            Returns:
+                bool: True if the device's role matches the specified roles, False otherwise.
+            """
             return dev_dict["device_role"]["slug"] in cli_opts.role
 
         filter_functions.append(filter_role)
 
     if cli_opts.exclude_role:
 
-        def filter_ex_role(dev_dict):
+        def filter_ex_role(dev_dict: dict) -> bool:
+            """Filter to exclude devices that match the specified roles.
+
+            Args:
+                dev_dict (dict): The device record to filter.
+
+            Returns:
+                bool: True if the device's role does not match the excluded roles, False otherwise.
+            """
             return dev_dict["device_role"]["slug"] not in cli_opts.exclude_role
 
         filter_functions.append(filter_ex_role)
 
     if cli_opts.exclude_tag:
         ex_tag_set = set(cli_opts.exclude_tag)
 
-        def filter_ex_tag(dev_dict):
+        def filter_ex_tag(dev_dict: dict) -> bool:
+            """Filter to exclude devices that have the specified tags.
+
+            Args:
+                dev_dict (dict): The device record to filter.
+
+            Returns:
+                bool: True if the device does not have any of the excluded tags, False otherwise.
+            """
             return not set(dev_dict["tags"]) & ex_tag_set
 
         filter_functions.append(filter_ex_tag)
 
-    def apply_filters():
-        for dev_dict in device_list:
-            if all(fn(dev_dict) for fn in filter_functions):
-                yield dev_dict
+    for dev_dict in device_list:
+        if all(fn(dev_dict) for fn in filter_functions):
+            yield dev_dict
 
-    return apply_filters() if filter_functions else iter(device_list)
 
+def build_inventory() -> None:
+    """Build the inventory by fetching device records and creating a CSV file.
 
-def build_inventory():
+    This function handles CLI parsing, inventory fetching, and CSV file creation.
+    """
     cli_opts = cli()
```

Comment:
To improve code modularity, consider separating the CLI parsing logic from the network request logic. [medium]